test/test.go

```diff
@@ -133,44 +133,68 @@ func FindOrCreateTest(s leakybucket.Storage) func(*testing.T) {
 // It is meant to be used by leakybucket implementers who wish to test this.
 func ThreadSafeAddTest(s leakybucket.Storage) func(*testing.T) {
 	return func(t *testing.T) {
-		// Make a bucket of size `n`. Spawn `n+1` goroutines that each try to take one token.
+		// Make a bucket of size `n`. Spawn `n+k` goroutines that each try to take one token.
 		// We should see the bucket transition through having `n-1`, `n-2`, ... 0 remaining capacity.
-		// We should also witness one error when the bucket has reached capacity.
+		// We should also witness k errors when the bucket has reached capacity.
 		n := 100
+		k := 50
 		bucket, err := s.Create("testbucket", uint(n), time.Minute)
 		if err != nil {
 			t.Fatal(err)
 		}
-		remaining := map[uint]bool{}     // record observed "remaining" counts. (ab)using map as set here
-		remainingMutex := sync.RWMutex{} // maps are not threadsafe
-		errors := []error{}              // record observed errors
-		var wg sync.WaitGroup
-		for i := 0; i < n+1; i++ {
-			wg.Add(1)
+
+		// process errors
+		var wgErrors sync.WaitGroup
+		errors := make(chan error)
+		wgErrors.Add(1)
+		go func() {
+			defer wgErrors.Done()
+			count := 0
+			for err := range errors {
+				count += 1
+				if err != leakybucket.ErrorFull {
+					t.Errorf("got an error that is not ErrorFull: %s", err)
+				}
+			}
+			if count != k {
+				t.Errorf("got %d errors, expected %d", count, k)
+			}
+		}()
+
+		// process remaining values returned from add()
+		var wgRemaining sync.WaitGroup
+		remaining := make(chan uint)
+		wgRemaining.Add(1)
+		go func() {
+			defer wgRemaining.Done()
+			count := 0
+			for _ = range remaining {
+				count += 1
+			}
+			if count != n {
```

Comment:
What is the value of this test? Seems like the only time this wouldn't be `n` is if more jobs succeeded than should have, which would already be caught by your assertion on the number of errors.